core/src/block.rs

```diff
@@ -420,7 +425,63 @@ mod valid {
             genesis_account: &AccountId,
             state_block: &mut StateBlock<'_>,
         ) -> WithEvents<Result<ValidBlock, (SignedBlock, BlockValidationError)>> {
-            let expected_block_height = state_block
+            if let Err(error) =
+                Self::validate_header(&block, topology, genesis_account, state_block)
+            {
+                return WithEvents::new(Err((block, error.into())));
+            }
+
+            if let Err(error) =
+                Self::validate_transactions(&block, expected_chain_id, genesis_account, state_block)
+            {
+                return WithEvents::new(Err((block, error.into())));
+            }
+
+            WithEvents::new(Ok(ValidBlock(block)))
+        }
+
+        /// Same as `validate` but:
+        /// * Block header will be validated with read-only state
+        /// * If block header is valid, `voting_block` will be released,
+        ///   and transactions will be validated with write state
+        pub fn validate_keep_voting_block<'state>(
```

Comment:
why do we still need `Validate::validate`?

Reply:
it seems to me that we could have just one method

Reply:
It is used in cases when block is not received from other peers, e.g. in `Sumeragi::init_commit_genesis` for genesis block (provided offline) or `SumeragiHandle::replay_block` (block read from kura)

Reply:
even in those cases, wouldn't you be able to just pass `None` for `voting_block`?

Reply:
> even in those cases, wouldn't you be able to just pass `None` for `voting_block`?

It will be not enough. These functions are also different in `StateBlock` handling. `validate` takes `StateBlock`, but `validate_keep_voting_block` takes `State` and additionally returns `StateBlock` in success case

Reply:
it looks to me as if you could send `State` and return `StateBlock` in those cases also?

Reply:
Probably yes, but it will complicate usual usages of `validate`, isn't it? Like `validate_keep_voting_block` is specialized function for specific use case thus have a bit more complicated API, and `validate` is general function with usual (easier to understand) API. There are no duplication between these functions (all common parts are extracted to `validate_header` and `validate_transactions`), so I am not sure what could be wrong with having two different functions

Reply:
my concern was that someone will use `validate` when they should have used `validate_keep_voting_block`